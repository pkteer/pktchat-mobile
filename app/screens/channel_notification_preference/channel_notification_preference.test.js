// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {ViewTypes} from '@constants';
import Preferences from '@mm-redux/constants/preferences';
import SectionItem from '@screens/settings/section_item';
import {shallowWithIntlMessages} from '@test/intl-test-helper';

import ChannelNotificationPreference from './channel_notification_preference';

function makeProps(pushNotificationLevel, pushThreadsNotificationLevel, isCollapsedThreadsEnabled = false) {
    return {
        actions: {
            updateChannelNotifyProps: jest.fn(),
        },
        channelId: 'channel_id',
        globalNotifyProps: {
            push: 'mention',
        },
        userId: 'user_id',
        notifyProps: {
            push: pushNotificationLevel,
            push_threads: pushThreadsNotificationLevel,
        },
        theme: Preferences.THEMES.denim,
        isCollapsedThreadsEnabled,
    };
}

function checkNotificationSelected(pushNotificationLevel, trueIdx) {
    const wrapper = shallowWithIntlMessages(
        <ChannelNotificationPreference
            {...makeProps(pushNotificationLevel)}
        />,
    );

    const sectionItems = wrapper.find(SectionItem);

    expect(sectionItems.exists()).toBe(true);
    expect(sectionItems.length).toBe(4);

    sectionItems.forEach((sectionItem, idx) => {
        expect(sectionItem.prop('selected')).toBe(idx === trueIdx);
    });
}

describe('ChannelNotificationPreference', () => {
    test('should match snapshot', () => {
        const baseProps = makeProps('default');
        const wrapper = shallowWithIntlMessages(
            <ChannelNotificationPreference {...baseProps}/>,
        );

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    test('should have correct setting selected', () => {
        checkNotificationSelected(null, 0);
        checkNotificationSelected(ViewTypes.NotificationLevels.DEFAULT, 0);
        checkNotificationSelected(ViewTypes.NotificationLevels.ALL, 1);
        checkNotificationSelected(ViewTypes.NotificationLevels.MENTION, 2);
        checkNotificationSelected(ViewTypes.NotificationLevels.NONE, 3);
    });

    test('should save on click, when CRT off', () => {
        const props = makeProps('default');
        const wrapper = shallowWithIntlMessages(
            <ChannelNotificationPreference {...props}/>,
        );

        // click on 'Never' -- last item
        wrapper.find(SectionItem).at(3).dive().simulate('press');

        expect(props.actions.updateChannelNotifyProps).toHaveBeenCalledTimes(1);
        expect(props.actions.updateChannelNotifyProps).toBeCalledWith(
            props.userId,
            props.channelId,
            {push: ViewTypes.NotificationLevels.NONE},
        );
    });

    test('should save on click, when CRT on', () => {
        const props = makeProps('default', undefined, true);
        const wrapper = shallowWithIntlMessages(
            <ChannelNotificationPreference {...props}/>,
        );

        // click on 'Never' -- last item
        wrapper.find(SectionItem).at(3).dive().simulate('press');

        expect(props.actions.updateChannelNotifyProps).toHaveBeenCalledTimes(1);
        expect(props.actions.updateChannelNotifyProps).toBeCalledWith(
            props.userId,
            props.channelId,
            {
                push: ViewTypes.NotificationLevels.NONE,
                push_threads: ViewTypes.NotificationLevels.ALL,
            },
        );
    });

    test('should show push_threads switch, when CRT is on and notifyLevel is mention', () => {
        const props = makeProps('all', undefined, true);
        const wrapper = shallowWithIntlMessages(
            <ChannelNotificationPreference {...props}/>,
        );

        expect(wrapper.find(SectionItem)).toHaveLength(4);

        wrapper.setState({
            notificationLevel: 'mention',
        });

        expect(wrapper.find(SectionItem)).toHaveLength(5);
    });

    test('should show push_threads switch, when CRT is off and notifyLevel is mention', () => {
        const props = makeProps('mention', undefined, false);
        const wrapper = shallowWithIntlMessages(
            <ChannelNotificationPreference {...props}/>,
        );
        expect(wrapper.find(SectionItem)).toHaveLength(4);
    });
});
