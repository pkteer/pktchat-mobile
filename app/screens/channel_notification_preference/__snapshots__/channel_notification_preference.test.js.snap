// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelNotificationPreference should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgba(63,67,80,0.03)",
      "flex": 1,
    }
  }
  testID="channel_notification_preference.screen"
>
  <Connect(StatusBar) />
  <RNCSafeAreaView
    edges={
      Array [
        "left",
        "right",
      ]
    }
  >
    <InjectIntl(FormattedText)
      defaultMessage="Send Notifications"
      id="channel_notifications.preference.header"
      style={
        Object {
          "backgroundColor": "rgba(63,67,80,0.03)",
          "color": "rgba(63,67,80,0.56)",
          "fontSize": 13,
          "padding": 16,
          "paddingTop": 26,
          "textTransform": "uppercase",
        }
      }
    />
  </RNCSafeAreaView>
  <ScrollView
    alwaysBounceVertical={false}
    contentContainerStyle={
      Object {
        "backgroundColor": "rgba(63,67,80,0.03)",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(255,255,255,1)",
        }
      }
    >
      <RNCSafeAreaView
        edges={
          Array [
            "left",
            "right",
          ]
        }
      >
        <View>
          <View
            style={
              Object {
                "backgroundColor": "rgba(63,67,80,0.1)",
                "height": 1,
                "marginLeft": 16,
                "width": "100%",
              }
            }
          />
          <sectionItem
            action={[Function]}
            actionType="select"
            actionValue="default"
            label={
              <InjectIntl(FormattedText)
                defaultMessage="Global default ({notifyLevel})"
                id="channel_notifications.preference.global_default"
                values={
                  Object {
                    "notifyLevel": "Mentions",
                  }
                }
              />
            }
            selected={true}
            theme={
              Object {
                "awayIndicator": "#ffbc1f",
                "buttonBg": "#1c58d9",
                "buttonColor": "#ffffff",
                "centerChannelBg": "#ffffff",
                "centerChannelColor": "#3f4350",
                "codeTheme": "github",
                "dndIndicator": "#d24b4e",
                "errorTextColor": "#d24b4e",
                "linkColor": "#386fe5",
                "mentionBg": "#ffffff",
                "mentionColor": "#1e325c",
                "mentionHighlightBg": "#ffd470",
                "mentionHighlightLink": "#1b1d22",
                "newMessageSeparator": "#cc8f00",
                "onlineIndicator": "#3db887",
                "sidebarBg": "#1e325c",
                "sidebarHeaderBg": "#192a4d",
                "sidebarHeaderTextColor": "#ffffff",
                "sidebarTeamBarBg": "#14213e",
                "sidebarText": "#ffffff",
                "sidebarTextActiveBorder": "#5d89ea",
                "sidebarTextActiveColor": "#ffffff",
                "sidebarTextHoverBg": "#28427b",
                "sidebarUnreadText": "#ffffff",
                "type": "Denim",
              }
            }
          />
        </View>
        <View>
          <View
            style={
              Object {
                "backgroundColor": "rgba(63,67,80,0.1)",
                "height": 1,
                "marginLeft": 16,
                "width": "100%",
              }
            }
          />
          <sectionItem
            action={[Function]}
            actionType="select"
            actionValue="all"
            label={
              <InjectIntl(FormattedText)
                defaultMessage="For all activity"
                id="channel_notifications.preference.all_activity"
              />
            }
            selected={false}
            theme={
              Object {
                "awayIndicator": "#ffbc1f",
                "buttonBg": "#1c58d9",
                "buttonColor": "#ffffff",
                "centerChannelBg": "#ffffff",
                "centerChannelColor": "#3f4350",
                "codeTheme": "github",
                "dndIndicator": "#d24b4e",
                "errorTextColor": "#d24b4e",
                "linkColor": "#386fe5",
                "mentionBg": "#ffffff",
                "mentionColor": "#1e325c",
                "mentionHighlightBg": "#ffd470",
                "mentionHighlightLink": "#1b1d22",
                "newMessageSeparator": "#cc8f00",
                "onlineIndicator": "#3db887",
                "sidebarBg": "#1e325c",
                "sidebarHeaderBg": "#192a4d",
                "sidebarHeaderTextColor": "#ffffff",
                "sidebarTeamBarBg": "#14213e",
                "sidebarText": "#ffffff",
                "sidebarTextActiveBorder": "#5d89ea",
                "sidebarTextActiveColor": "#ffffff",
                "sidebarTextHoverBg": "#28427b",
                "sidebarUnreadText": "#ffffff",
                "type": "Denim",
              }
            }
          />
        </View>
        <View>
          <View
            style={
              Object {
                "backgroundColor": "rgba(63,67,80,0.1)",
                "height": 1,
                "marginLeft": 16,
                "width": "100%",
              }
            }
          />
          <sectionItem
            action={[Function]}
            actionType="select"
            actionValue="mention"
            label={
              <InjectIntl(FormattedText)
                defaultMessage="Only mentions and direct messages"
                id="channel_notifications.preference.only_mentions"
              />
            }
            selected={false}
            theme={
              Object {
                "awayIndicator": "#ffbc1f",
                "buttonBg": "#1c58d9",
                "buttonColor": "#ffffff",
                "centerChannelBg": "#ffffff",
                "centerChannelColor": "#3f4350",
                "codeTheme": "github",
                "dndIndicator": "#d24b4e",
                "errorTextColor": "#d24b4e",
                "linkColor": "#386fe5",
                "mentionBg": "#ffffff",
                "mentionColor": "#1e325c",
                "mentionHighlightBg": "#ffd470",
                "mentionHighlightLink": "#1b1d22",
                "newMessageSeparator": "#cc8f00",
                "onlineIndicator": "#3db887",
                "sidebarBg": "#1e325c",
                "sidebarHeaderBg": "#192a4d",
                "sidebarHeaderTextColor": "#ffffff",
                "sidebarTeamBarBg": "#14213e",
                "sidebarText": "#ffffff",
                "sidebarTextActiveBorder": "#5d89ea",
                "sidebarTextActiveColor": "#ffffff",
                "sidebarTextHoverBg": "#28427b",
                "sidebarUnreadText": "#ffffff",
                "type": "Denim",
              }
            }
          />
        </View>
        <View>
          <View
            style={
              Object {
                "backgroundColor": "rgba(63,67,80,0.1)",
                "height": 1,
                "marginLeft": 16,
                "width": "100%",
              }
            }
          />
          <sectionItem
            action={[Function]}
            actionType="select"
            actionValue="none"
            label={
              <InjectIntl(FormattedText)
                defaultMessage="Never"
                id="channel_notifications.preference.never"
              />
            }
            selected={false}
            theme={
              Object {
                "awayIndicator": "#ffbc1f",
                "buttonBg": "#1c58d9",
                "buttonColor": "#ffffff",
                "centerChannelBg": "#ffffff",
                "centerChannelColor": "#3f4350",
                "codeTheme": "github",
                "dndIndicator": "#d24b4e",
                "errorTextColor": "#d24b4e",
                "linkColor": "#386fe5",
                "mentionBg": "#ffffff",
                "mentionColor": "#1e325c",
                "mentionHighlightBg": "#ffd470",
                "mentionHighlightLink": "#1b1d22",
                "newMessageSeparator": "#cc8f00",
                "onlineIndicator": "#3db887",
                "sidebarBg": "#1e325c",
                "sidebarHeaderBg": "#192a4d",
                "sidebarHeaderTextColor": "#ffffff",
                "sidebarTeamBarBg": "#14213e",
                "sidebarText": "#ffffff",
                "sidebarTextActiveBorder": "#5d89ea",
                "sidebarTextActiveColor": "#ffffff",
                "sidebarTextHoverBg": "#28427b",
                "sidebarUnreadText": "#ffffff",
                "type": "Denim",
              }
            }
          />
        </View>
      </RNCSafeAreaView>
      <View
        style={
          Object {
            "backgroundColor": "rgba(63,67,80,0.1)",
            "height": 1,
            "marginLeft": 16,
            "width": "100%",
          }
        }
      />
    </View>
  </ScrollView>
</View>
`;
